src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +492,355 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+ *
+ * Verifies whether the uptane client is not at pending state after installation finalization failure
+ *
+ * Checks actions:
+ *
+ * - [x] Update is not in pending state anymore after failed finalization
+ */
+TEST(Aktualizr, FinalizationFailure) {
+  TemporaryDirectory temp_dir;
+  auto http_server_mock = std::make_shared<HttpInstallationFailed>(temp_dir.Path());
+  Config conf = UptaneTestCommon::makeTestConfig(temp_dir, http_server_mock->tls_server);
+  conf.pacman.fake_need_reboot = true;
+  conf.uptane.force_install_completion = true;
+  conf.uptane.polling_sec = 0;
+  conf.bootloader.reboot_sentinel_dir = temp_dir.Path();
+  auto storage = INvStorage::newStorage(conf.storage);
+
+  std::vector<std::string> expected_event_order = {
+      "SendDeviceDataComplete", "UpdateCheckComplete",    "DownloadProgressReport", "DownloadTargetComplete",
+      "DownloadProgressReport", "DownloadTargetComplete", "AllDownloadsComplete",   "InstallStarted",
+      "InstallTargetComplete",  "InstallStarted",         "InstallTargetComplete",  "AllInstallsComplete"};
+
+  std::vector<std::string> expected_report_order = {
+      "EcuDownloadStarted",     "EcuDownloadCompleted",   "EcuDownloadStarted",     "EcuDownloadCompleted",
+      "EcuInstallationStarted", "EcuInstallationApplied", "EcuInstallationStarted", "EcuInstallationCompleted"};
+
+  const std::string primary_ecu_id = "CA:FE:A6:D2:84:9D";
+  const std::string secondary_ecu_id = "secondary_ecu_serial";
+
+  {
+    // get update, install them and emulate reboot
+    Aktualizr aktualizr(conf, storage, http_server_mock);
+    EventHandler event_hdlr(aktualizr);
+    aktualizr.Initialize();
+
+    // verify currently installed version
+    std::vector<Uptane::Target> installed_versions;
+    size_t current_version{0};
+    size_t pending_version{0};
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+
+    // for some reason there is no any installed version at initial Aktualizr boot/run
+    // IMHO it should return currently installed version
+    ASSERT_TRUE(installed_versions.empty());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    auto aktualizr_cycle_thread = aktualizr.RunForever();
+    auto aktualizr_cycle_thread_status = aktualizr_cycle_thread.wait_for(std::chrono::seconds(10));
+
+    ASSERT_EQ(aktualizr_cycle_thread_status, std::future_status::ready);
+    ASSERT_TRUE(aktualizr.uptane_client_->bootloader->rebootDetected());
+    ASSERT_TRUE(event_hdlr.checkRecievedEvents(expected_event_order));
+    ASSERT_TRUE(aktualizr.uptane_client_->hasPendingUpdates());
+    ASSERT_TRUE(http_server_mock->checkRecievedReports(expected_report_order));
+    // Aktualizr reports to a server that installation was successfull in spite the fact that
+    // the installation requires `finalization`
+    ASSERT_TRUE(http_server_mock->wasInstallSuccessfull());
```

Comment:
Even with your fix? That seems like a bug.

Reply:
It's actually a bug in my test. Aktualizr reports about both ECUs and installation for the secondary ECU succeeds.